### PR TITLE
Allow creating AnsibleModule from DOCUMENTATION

### PR DESCRIPTION
**Pull request by emonty**
##### SUMMARY

As a module author it is currently necessary to list all of the module's
arguments twice. Once in the DOCUMENTATION string and once in the
argument_spec argument to AnsibleModule. The data is essentially the
same, with the exception of the descriptions in the docs, and the no_log
parameter which isn't accounted for in the docs.

CI jobs exit now to ensure that modules have properly validated DOCUMENTATION.

Since the DOCUMENTATION is required and is required to be correctly
parsable yaml, it's not terribly hard to parse it server-side when we
read the module and then to send it in the json we send over the wire
with the parameters. Doing so invocation-side avoids needing yaml or the
AST parsing bits on the remote end.

Once it's passed over the wire, the AnsibleModule constructor can allow
argument_spec to be empty, and if so fill it in with the data passed
received. This is obviously optional, as breaking AnsibleModule
would be double-plus bad, so explicitly provided argument_spec wins over
a received argument spec.

The docsite generation tool and the ansible-doc command-line tool have
been updated to understand no_log and removed_in_version in the docs and
to emit text indicating the effects of both options.

As a next step we could add support for specifying:

  * mutually_exclusive
  * required_together
  * required_one_of
  * required_if
  * supports_check_mode

in the DOCUMENTATION and reading them back out in AnsibleModule. Those
items are currently not exposed in the documentation as it is, so even
just being able to document them would be a win.

add_file_common_args is already handled because documentation fragements
are processed by the doc parser. As long as a module that does
add_file_common_args=True lists extends_documentation_fragment: files
in its DOCUMENTATION section no special support is needed.

Once we add support to the DOCUMENTATION for the additional flags above,
there will no longer be things one can specify in AnsibleModule constructor
args that cannot be specified in a DOCUMENTATION block. It should be possible
for rich and complex modules to simply do:

  module = AnsibleModule()

and work properly if they have correct and well-formed DOCUMENTATION.

Finally, having a parsed version of the DOCUMENTATION available at
module execution time means it's possible to do parameter validation
before making remote connections at all. Doing that may be a terrible
idea though, so it's also left as a follow-up idea.

##### ISSUE TYPE
 - Feature Pull Request

### CODE REVIEW

**Comment by abadger** [comment]
We've been talking about ways to make the argument spec available
controller side so it's good to see an implementation.  However, this one
points out some problems that will need to be thought about for whatever
our final implementation is.

Ast parsing is slow.  So we'll have to figure it how to cache that for each
module.

We don't want to mix the variable parameters with the invariant code.  We
should place the invariants into the same container as the module itself
(solving this may solve the ast caching issue)

We're redesigning ansible module for either 2.5 or 2.6 (redesign under way;
implementation may not be until 2.6).  Will need to make sure this works
with that.

Perhaps this can be a feature of the redesign only?  @nitzmahone, how would
this fit in with a redesigned documentation?  Is this something we want to
do at the same time as redesigning ansible module or is that too much churn
at the same time?

**Comment by bcoca** [comment]
I had been thinking in similar manner, see a early draft https://github.com/bcoca/ansible/tree/doc2args

@abadger since i'm already doing something very similar to other plugins, i had at one point 'cached' the options by attaching them as plugin properties, so the first time they are loaded would pay the cost for the rest of the run. I ended up just putting the information into the configdata object ... which can still easily accommodate modules (it is namespaced).

**Comment by abadger** [comment]
<nod>. But what we really want is to write the assembled module to disk
just once per run.

So pushing the module arg spec into the zip file somewhere.

**Comment by mattclay** [comment]
CI failure in unit and integration tests: https://app.shippable.com/github/ansible/ansible/runs/36065

**Comment by dharmabumstead** [comment]
Closing due to inactivity. Please feel free to reopen and rebase for reconsideration. Thanks!
